### PR TITLE
Add FxCop Crypto rules

### PR DESCRIPTION
**Pull request by qinxgit**
Added FxCop Crypto rules and Xml rules, as well as corresponding tests. @genlu 
@mavasani @srivatsn 
 @sostanko

# CA5350 Do Not Use Weak Cryptographic Algorithms 
## Cause 
Cryptographic algorithms such as 3DES, SHA1, RIPEMD160 and Rijindael can be easily subjected to brute force attacks, and therefore, are considered to be weak. 
 
## Rule Description 
Weak cryptographic algorithms are used today for a number of reasons, but they should not guarantee the confidentiality of the data they protect. 
 
Today, technological advancements in computational power make it feasible for attackers to brute force their way into the secrets these cryptographic algorithms keep.  
 
The rule triggers once it finds 3DES, SHA1, RIPEMD160 or Rijindael algorithms in the code and throws a warning to the user. 
 
## How to Fix Violations 
Use cryptographically stronger options: 
For TripleDES or Rijindael encryption algorithms, use AES encryption 
For SHA1 or RIPEMD160 hashing functions, use ones in the SHA-2 family (e.g. SHA512, SHA384, SHA256) 
 ## When to Suppress Warnings 
Suppress a warning from this rule when the level of protection needed for the data does not require a security guarantee. 

# CA5351 Do Not Use Broken Cryptographic Algorithms 

## Cause 
Cryptographic algorithms such as MD5, DES, RC2 and DSA expose unnecessary risk and may result in the exposure of sensitive information through trivial attacking techniques, such as brute force attacks and hashing collisions.  
 
## Rule Description 
Broken algorithms are considered insecure and their use should be discouraged. MD5 hashing is susceptible to computationally feasible hashing collisions, which allows the attacker to try billions of password combinations per second on a single processor to find unique input combinations that results in the same hashing output. DES encryption contains a small key size, which can be broken in less than a day. RC2 encryption is susceptible to a related-key attack, where the attacker finds mathematical relationships between all key values. DSA encryption can be broken when number generation does not output unique values for its k parameter.  
 
The rule triggers once it finds any of the above cryptographic functions in the code and throws a warning to the user. 
 
## How to Fix Violations 
Use cryptographically stronger options:  
For MD5, use hashes in the SHA-2 family (e.g. SHA512, SHA384, SHA256) 
For DES, RC2 and DSA, use AES encryption 
  
## When to Suppress Warnings 
Do not suppress a warning from this rule.

### CODE REVIEW

**Comment by qinxgit** [comment]
@genlu per your suggestion, reduce the pull request to crypto rules first.

**Comment by genlu** [comment]
FYI @michaelcfanning @tmeschter

File: src/FxCop/Desktop.Analyzers/CSharp/CSharpDesktopAnalyzers.csproj [hunk 1]
```diff
@@ -100,8 +100,9 @@
   </ItemGroup>
   <ItemGroup />
   <ItemGroup>
+    <Compile Include="Security\DoNotUseInsecureCryptographicAlgorithms.cs" />
     <Compile Include="Shared\SyntaxNodeHelper.cs" />
-    <Compile Include="Security\DoNotCatchCorruptedStateExceptionsAnalyzer.cs" />
+    <Compile Include="Security\DoNotCatchCorruptedStateExceptions.cs" />
```
Comment (tmeschter) [inline src/FxCop/Desktop.Analyzers/CSharp/CSharpDesktopAnalyzers.csproj]: Why the rename? Typically file names reflect the name of the type within.

File: src/FxCop/Desktop.Analyzers/Core/DesktopAnalyzers.csproj [hunk 2]
```diff
@@ -93,9 +93,10 @@
     </Content>
   </ItemGroup>
   <ItemGroup>
+    <Compile Include="Security\DoNotUseInsecureCryptographicAlgorithms.cs" />
     <Compile Include="Security\SecurityTypes.cs" />
     <Compile Include="Security\CompilationSecurityTypes.cs" />
-    <Compile Include="Security\DoNotCatchCorruptedStateExceptionsAnalyzer.cs" />
+    <Compile Include="Security\DoNotCatchCorruptedStateExceptions.cs" />
```
Comment (tmeschter) [inline src/FxCop/Desktop.Analyzers/Core/DesktopAnalyzers.csproj]: Again, the name change seems superfluous and does not match the existing style.

File: src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        private static readonly LocalizableString s_localizableDoNotUseRIPEMD160Title = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRIPEMD160));
+        private static readonly LocalizableString s_localizableDoNotUseRIPEMD160Description = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRIPEMD160Description));
+        private static readonly LocalizableString s_localizableDoNotUseDSATitle = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseDSA));
+        private static readonly LocalizableString s_localizableDoNotUseDSADescription = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseDSADescription));
+        private static readonly LocalizableString s_localizableDoNotUseRijndaelTitle = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRijndael));
+        private static readonly LocalizableString s_localizableDoNotUseRijndaelDescription = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRijndaelDescription));
+
+
```
Comment (tmeschter) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs]: Extraneous whitespace.

File: src/FxCop/Desktop.Analyzers/CSharp/CSharpDesktopAnalyzers.csproj (hunk 1 above)
Comment (qinxgit) [inline src/FxCop/Desktop.Analyzers/CSharp/CSharpDesktopAnalyzers.csproj]: Looks like this one and the one  above "DoNotUseInsecureCryptographicAlgorithms" both didn't use the class names as file names. I believe this should be fixed. 

@genlu 
What do you think?

File: src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.RegisterCompilationStartAction(
+                (context) =>
+                {
+                    var cryptTypes = new CompilationSecurityTypes(context.Compilation);
+                    if (ReferencesAnyTargetType(cryptTypes))
+                    {
+                        Analyzer analyzer = GetAnalyzer(context, cryptTypes);
```
Comment (tmeschter) [inline src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs]: Why does `GetAnalyzer` return an analyzer?

Comment (qinxgit) [inline src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs]: I believe this GetAnalyzer is an abstract method which will get either VB analyzer or CSharp analyzer at runtime, according to what language it is analyzing.

File: src/Roslyn/Core/ApiDesign/DeclarePublicAPIFix.cs [hunk 5]
```diff
@@ -129,6 +129,14 @@ protected override Task<Solution> GetChangedSolutionAsync(CancellationToken canc
...
                 return _createChangedAdditionalDocument(cancellationToken);
             }
+
+            public override string EquivalenceKey
+            {
+                get
+                {
+                    return nameof(AdditionalDocumentChangeAction);
+                }
+            }
```
Comment (tmeschter) [inline src/Roslyn/Core/ApiDesign/DeclarePublicAPIFix.cs]: This seems to be unrelated.

File: src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs (hunk 4 above)
Comment (tmeschter) [inline src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs]: Yes, but why? The callers of this method don't do anything with the value returned.

**Comment by tmeschter** [comment]
This change adds a lot of very similar strings. I can understand that for backwards compatibility the title strings can't really change, but would it make sense to combine the message strings, perhaps?

File: src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs (hunk 4 above)
Comment (qinxgit) [inline src/FxCop/System.Security.Cryptography.Hashing.Algorithms.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs]: I will remove the receiving variable of that.

Searched and find in some cases, the analyzer will be used in this statement, in our CA3077 rule, which will create a request later:

Analyzer analyzer = GetAnalyzer(context, xmlTypes, version);
context.RegisterSymbolAction(analyzer.AnalyzeSymbol, SymbolKind.NamedType);

This statement appears to be in both CSharp and VB cases, so I guess it get moved into the base class of CA3077.

File: src/Roslyn/Core/ApiDesign/DeclarePublicAPIFix.cs (hunk 5 above)
Comment (qinxgit) [inline src/Roslyn/Core/ApiDesign/DeclarePublicAPIFix.cs]: Yes. This is addressing a Roslyn rule RS1011. "... has the default value of 'null' for property 'EquivalenceKey'. Either override this property on '...' to return a non-null and unique value across all code actions per-fixer or use such an existing code action.	

Will remove it for now since it is not related to the purpose of this pull request.

File: src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs (hunk 3 above)
Comment (qinxgit) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithms.cs]: I believe this is due to the display of github. Here is it displayed on my desktop.
Image of code

**Comment by qinxgit** [comment]
To the comment about message strings, each different string is having a different scenario, it would be accurate and more actionable for the user. Of course the variable name of the string looks similar in code. Any suggestions to combine?

**Comment by qinxgit** [comment]
To the comment about message strings, each different string is having a different scenario, it would be accurate and more actionable for the user. Of course the variable name of the string looks similar in code. Any suggestions to combine?

File: src/FxCop/Desktop.Analyzers/CSharp/CSharpDesktopAnalyzers.csproj (hunk 1 above)
Comment (genlu) [inline src/FxCop/Desktop.Analyzers/CSharp/CSharpDesktopAnalyzers.csproj]: I agree that file name should be consistent with type name. That being said, it seems this is a problem for most of if not all analyzers under FxCop directory. I will address this issue with the clean-up/re-organization of the repo later.

File: src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using Desktop.Analyzers.Common;
+
+
+namespace Desktop.Analyzers
+{
+    public abstract class DoNotUseInsecureCryptographicAlgorithmsAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string DoNotUseMD5RuleId = "CA5350";
```
Comment (michaelcfanning) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs]: At one time, we'd discussed collapsing these checks into a single ID for a 'DoNotUseCryptographicallyWeakAlgorithms'. I think there's still merit to this position. @genlu, maybe we can discuss offline?

Comment (qinxgit) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs]: I updated code according to the feedback except michaelcfanning's one. I don't know what the next step will look like. So am I now waiting for your decisions? @genlu @michaelcfanning

Comment (genlu) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs]: @michaelcfanning I actually did implement the analyzer with a single ID, but then the team made a decision to keep using old IDs instead. One of the reasons, IIRC, is some algorithms is less broken than others and we want to give users the ability to choose.

Comment (qinxgit) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs]: So … what’s next step?

**Comment by tmeschter** [comment]
Why are there two abstract types named DoNotUseInsecureCryptographicAlgorithmsAnalyzer?

-Tom

**Comment by tmeschter** [comment]
In general this pull request would benefit from an expanded description explaining exactly which rules are being added, what they check for, why they are being added now, and why the code is organized the way it is.

In general, err on the side of including more explanation in a commit or pull request message, rather than less. An empty or perfunctory message assumes that the commit/pull request is entirely self-explanatory to anyone reading the code.

The reason I haven't signed off is that while the individual code changes make sense, I can't see the "big picture".

File: src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs (hunk 6 above)
Comment (michaelcfanning) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs]: I'm not sure the distinctions around algorithms are very helpful. We only require granularity in rule id for cases where there might be some mingling of issues that users want to keep separate (suppressing some, not suppressing others). So, if you applied a single exclusion to a method with a common ID, and you happened to intermix two cryptographically weak algorithms in that method, you would suppress violations for both. I don't think this intermixing is common. With Roslyn, you always have the option of using a #pragma to isolate issues (unless this check requires wrapping an entire type).

Long story short, I don't find the argument to create per algorithm ids compelling. Another thing to think about is the evolution of this check. New crypto algorithms may be added to the list (you know, when quantum computers crack all our existing <g>), it'd be nice not to have to author new rules. Any other thoughts out there?

**Comment by qinxgit** [comment]
@tmeschter To the question about 2 abstract types, I believe the reason behind is prepare this for the future .NET CLR 5. The one under System.Security.Cryptography.Hashing.Algorithms.Analyzers is for .NET CLR , and the other one under Desktop.Analyzers is for .NET Core. @genlu 

I will try to add the overall introduction to the pull request after I am done.

**Comment by qinxgit** [comment]
@michaelcfanning @genlu 
I combined the rule id together as we discussed in mails. See the latest commits. I didn't rename the class DoNotUseInsecureCryptographicAlgorithms 'cause the class will contain both Do Not Use Broken Cryptographic Algorithms and Do Not Use Weak Cryptographic Algorithms, so I think the current name DoNotUseInsecureCryptographicAlgorithms is appropriate.

**Comment by qinxgit** [comment]
@tmeschter  Added more details in the pull request description

**Comment by qinxgit** [comment]
@michaelcfanning @genlu @tmeschter 

A kind reminder to you guys to take a look at the new code.

File: src/FxCop/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx [hunk 7]
```diff
@@ -126,4 +126,46 @@
   <data name="DoNotCatchCorruptedStateExceptionsMessage" xml:space="preserve">
     <value>Modify {0} to catch and handle a more specific set of exception type(s) than {1} or rethrow the exception. </value>
   </data>
+  <data name="DoNotUseDES" xml:space="preserve">
+    <value>Do not use insecure cryptographic algorithm DES.</value>
+  </data>
```
Comment (genlu) [inline src/FxCop/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: Since we have two rules now, I think the title and description should be changed to reflect the change, RE: weak vs. broken algorithms, as well.

Comment (qinxgit) [inline src/FxCop/Desktop.Analyzers/Core/DesktopAnalyzersResources.resx]: @genlu  We are doing a complete overhaul of all strings. See document below. We are still working on it. Can I separate the string changes into later pull requests?


https://microsoft.sharepoint.com/teams/CESecEngineering/SecDevTools/Shared%20Documents/Specs,%20Functional%20and%20Design/Roslyn/Roslyn_UserFacingStrings.xlsx?web=1

**Comment by genlu** [comment]
LGTM assuming title/description strings will be updated in upcoming PRs.

**Comment by qinxgit** [comment]
@michaelcfanning Did you get a chance to review?

**Comment by qinxgit** [comment]
@michaelcfanning Does everything look OK now?

**Comment by genlu** [comment]
@michaelcfanning I'm merging this so I can use it for the project generation. We will address any further comments in subsequent PRs.
@qinxgit Could you please create the PR for strings? Thanks

File: src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        private static readonly LocalizableString s_localizableDoNotUseRIPEMD160Description = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRIPEMD160Description));
+        private static readonly LocalizableString s_localizableDoNotUseDSATitle = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseDSA));
+        private static readonly LocalizableString s_localizableDoNotUseDSADescription = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseDSADescription));
+        private static readonly LocalizableString s_localizableDoNotUseRijndaelTitle = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRijndael));
+        private static readonly LocalizableString s_localizableDoNotUseRijndaelDescription = DiagnosticHelpers.GetLocalizableResourceString(nameof(DesktopAnalyzersResources.DoNotUseRijndaelDescription));
+
+
+        internal static DiagnosticDescriptor DoNotUseMD5SpecificRule = CreateDiagnosticDescriptor(DoNotUseBrokenCryptographicRuleId,
```
Comment (srivatsn) [inline src/FxCop/Desktop.Analyzers/Core/Security/DoNotUseInsecureCryptographicAlgorithmsAnalyzer.cs]: We shouldn't have multiple descriptors with the same id. If they all have different titles then the configuration experience is very bad because only one of these will show up in the ruleset editor\solution explorer. We should have a generic title\description. The individual message can have a fillin that says which algorithm is being used and shouldn't be.
